File: 0953.验证外星语词典/0953-验证外星语词典.cpp
class Solution {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        unordered_map<char, int> dict;
        for( int i=0; i< order.size(); i++ ) {
            dict[order[i]] = i;
        }

        int maxLen = 0;
        for( int i=0; i< words.size(); i++ ) {
            maxLen = max(maxLen, (int)words[i].length());
        }

        for( int i=1 ;i<words.size(); i++ ) {
            for( int j=0; j<words[i-1].length(); j++ ) {
                int a = dict[words[i-1][j]];
                int b = j<words[i].length()?dict[words[i][j]]:-1;
                if( a < b ) {
                    break;
                }
                if( a > b ) {
                    return false;
                }
            }
        }

        return true;
    }
};

File: 0953.验证外星语词典/0953-验证外星语词典.go
func isAlienSorted(words []string, order string) bool {
    m := make(map[byte]int)
    for i:=0; i<len(order); i++ {
        m[order[i]] = i
    }

    for j:=1; j<len(words); j++ {
        for i:=0; i<len(words[j-1]); i++  {
            a := m[words[j-1][i]]
            b := -1
            if i < len(words[j])  {
                b = m[words[j][i]]
            }

            if a < b  {
                break
            }
            if a >b  {
                return false
            }
        }
    }

    return true
}